controller/lib/hal/adc.cpp

```diff
@@ -0,0 +1,240 @@
+ * easier to read spreadsheet is availabe here:
+ * https://docs.google.com/spreadsheets/d/1JOSQKxkQxXJ6MCMDI9PwUQ6kiuGdujR4D6EJN9u2LWg/edit#gid=0
+ *
+ * PA0 (ADC1_IN5)  - vin (not currently used)
+ * PA1 (ADC1_IN6)  - pressure
+ * PA4 (ADC1_IN9)  - inhale flow
+ * PB0 (ADC1_IN15) - exhale flow
+ *
+ * See chapter 16 of the reference manual
+ *
+ *****************************************************************/
+
+// How long a period (in seconds) we want to average the A/D readings.
+static constexpr float sample_history_time_sec = 0.001f;
+
+// Total number of A/D inputs we're sampling
+static constexpr int adc_channels = 3;
+
+// This constant controls how many times we have the A/D sample each input
+// and sum them before moving on to the next input.  The constant is set
+// as a log base 2, so a value of 3 for example would mean sample 8 times
+// (2^3 == 8).  Legal values range from 0 to 8.
+static constexpr int oversample_log2 = 4;
+static constexpr int oversample_count = 1 << oversample_log2;
+
+// This calculated constant gives the maximum A/D reading based on the
+// number of samples.
+static constexpr int max_adc_reading =
+    (oversample_log2 >= 4) ? 65536 : (1 << (12 + oversample_log2));
+
+// A/D sample time in CPU clock cycles.  Fixed for now
+// This is the time we give the analog input to charge the A/D sampling cap.
+static constexpr int adc_samp_time = 92;
+
+// Time of an A/D conversion in CPU clock cycles.  This is the sample time + 13
+static constexpr int adc_conversion_time = adc_samp_time + 13;
+
+// Calculate how long our history buffer needs to be based on the above.
+static constexpr int adc_samp_history =
+    static_cast<int>(sample_history_time_sec * CPU_FREQ / adc_conversion_time /
+                     oversample_count / adc_channels);
+
+// This scaler converts the sum of the A/D readings (a total of
+// adc_samp_history) into a voltage.  The A/D is scaled so a value of 0
+// corresponds to 0 volts, and max_adc_reading corresponds to 3.3V
+static constexpr float adc_scaler = 3.3f / (max_adc_reading * adc_samp_history);
+
+// This buffer will hold the readings from the A/D
+static uint16_t adc_buff[adc_samp_history * adc_channels];
+
+void HalApi::InitADC() {
+
+  // Enable the clock to the A/D converter
+  EnableClock(ADC_BASE);
+
+  // Configure the 4 pins used as analog inputs
+  GPIO_PinMode(GPIO_A_BASE, 0, GPIO_PinMode::ANALOG);
+  GPIO_PinMode(GPIO_A_BASE, 1, GPIO_PinMode::ANALOG);
+  GPIO_PinMode(GPIO_A_BASE, 4, GPIO_PinMode::ANALOG);
+  GPIO_PinMode(GPIO_B_BASE, 0, GPIO_PinMode::ANALOG);
+
+  // Perform a power-up and calibration sequence on
+  // the A/D converter
+  ADC_Regs *adc = ADC_BASE;
+
+  // Exit deep power down mode and turn on the
+  // internal voltage regulator.
+  adc->adc[0].ctrl = 0x10000000;
+
+  // Wait for the startup time specified in the datasheet
+  // for the voltage regulator to become ready.
+  // The time in the datasheet is 20 microseconds, but
```

Comment:
Move comment from sensors.cpp that the 20ms number comes from a datasheet for a different part?

Reply:
uC data sheet, not sensor.  I'll make that more clear